Implemented more builtins

JasonTruter opened the pull request:
You can find references to the following builtins here:
https://www.openpolicyagent.org/docs/latest/policy-reference/

Added the following builtins:

- Numbers
- Aggregates
- Arrays
- Strings
- Regex
- Types
- Conversions

Note: Not all builtins in the reference guide have been added.

Also added Jest Testing framework

JasonTruter commented:
You might also need to remove my console.log for builtins. Forgot about that.

tsandall commented:
@patrick-east could you take a look?

patrick-east commented on src/opa.js:
```diff
@@ -117,7 +161,7 @@ async function _load_policy(policy_wasm, memory) {
 
     const builtins = _dumpJSON(wasm.instance, memory, wasm.instance.exports.builtins());
-
+    console.log({builtins})
```
I think you already mentioned it, but we'd probably want to remove the log message.

JasonTruter replied:
Fixed

patrick-east commented on src/opa.js:
```diff
@@ -49,13 +48,58 @@ function _dumpJSON(wasmInstance, memory, addr) {
 const builtinFuncs = {
-    plus: builtinPlus,
+    // Numbers
+    plus: (a, b) =>  a + b,
```
Just thinking outloud, this might be a good opportunity to consider the file/directory structure for the SDK. It might make sense to move these out into a separate directory and then sub files for the groups. Something like:
```
./src/builtins
./src/builtins/numbers.js
./src/builtins/aggregates.js
./src/builtins/arrays.js
.
.
.
```

The reasoning being that over time these are likely to not be so simple, and would likely benefit from being more standalone functions with room for docs and stuff.

If we do this we'd probably want to have some sort of registration step (similar to how they are done in OPA itself where each file of builtins handles registering themselves with the builtin name).

Thoughts?

JasonTruter replied:
Probably begin by breaking up the functions without registration of a builtin for now, I'll import into opa from the builtins.

patrick-east commented on src/opa.js:
```diff
@@ -49,13 +48,58 @@ function _dumpJSON(wasmInstance, memory, addr) {
+    "array.slice": (arr, startIndex, stopIndex) => arr.slice(startIndex, stopIndex),
+
+    // Strings
+    contains: (string, search) => string.includes(search),
```
Thinking through some of this I wonder how strict we need to be about type checking. In the builtin implementations for OPA in Golang they pretty much all use a set of shared helpers to get operands and verify the type is what was expected (eg, passing a number or array into this one as `string` would be caught as invalid and an error raised).

My main concern is that we should have consistent behavior for builtin functions between normal OPA, Golang WASM runtime w/ OPA binary, and JS WASM runtime w/ OPA binary, etc. The same policy+data+query+input should give the same result on all of them.

patrick-east commented:
@JasonTruter thanks for submitting this PR! The changes look good and I'm excited to see more usage of the library! I added a few comments but I don't think many are really blocking. Most of it we can improve/refactor over time as needed.

The main higher level ask would be if we had some way to test these. Do you have any experience working with test frameworks for these types of things? We don't have anything in place right now so its a good opportunity to try something out 😄

Jason-ikhokha commented:
@patrick-east Yes ofcourse, it would be great if we could use jest to test these functions and I definately think that it would be good to seperate them into seperate files. I'm also excited to start contributing to this SDK as we currently use it to evaluate policies in AWS.

patrick-east commented:
Jest looks good to me. Logistics-wise would you be ok with adding in some (maybe minimal) integration of it with this PR? I'm thinking enough to get to where we have _some_ testing framework in place and tests for some of these so that we can build off it going forward.

JasonTruter commented:
Thanks guys, I'm going to begin working on this.

JasonTruter commented:
Going to split builtins into builtins folder. Thereafter import the builtins folder into opa.js and map each builtin to the builtin name on the map manually. I will begin by writing tests for some of the builtin also.

patrick-east commented:
@JasonTruter keep an eye out for the (growing) list of functions being implemented in C and baked into the WASM binary: https://github.com/open-policy-agent/opa/blob/master/internal/planner/planner.go#L26

It will likely reduce the amount of builtin functions we need to do in javascript quite a bit

JasonTruter commented:
> @JasonTruter keep an eye out for the (growing) list of functions being implemented in C and baked into the WASM binary: https://github.com/open-policy-agent/opa/blob/master/internal/planner/planner.go#L26
> 
> It will likely reduce the amount of builtin functions we need to do in javascript quite a bit

Cool will do, @patrick-east please review latest changes.

JasonTruter commented:
> @patrick-east you can't use the `string` keyword as argument name. for example:
> 
> ```
> split: (string, delimiter) => string.split(delimiter)
> ```
> 
> should be
> 
> ```
> split: (s, delimiter) => s.split(delimiter)
> ```
> 
> or else it breaks

I've fixed those variable names

JasonTruter commented:
Might also be a good idea to add a prettier. config so that we standardize formatting

JasonTruter commented:
Cool will be assisting with contributions to npm opa wasm. Thanks for the
merge

On Wed, 13 May 2020, 04:41 Patrick East, <notifications@github.com> wrote:

> Merged #11 <https://github.com/open-policy-agent/npm-opa-wasm/pull/11>
> into master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-policy-agent/npm-opa-wasm/pull/11#event-3330070869>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOP2H7D5KLFOPHUDI6VDL63RRICFNANCNFSM4KXFLU2Q>
> .
>
